### PR TITLE
Various improvements to FloatCore

### PR DESCRIPTION
**Pull request by cuviper**
- New macros simplify forwarding method implementations.
  - `Float` and `Real` use this to compact their implementations.
  - `FloatCore` now forwards `std` implementations when possible.
- `FloatCore` now requires `NumCast`, like `Float does.
- New additions to `FloatCore`:
  - Constants like `min_value()` -> `f64::MIN`
  - Rounding methods `floor`, `ceil`, `round`, `trunc`, `fract`
  - `integer_decode` matching `Float`'s
- Fix NAN sign handling in `FloatCore` (rust-num/num#312, rust-lang/rust#42425)
- Fix overflow in `FloatCore::powi` exponent negation.
- Add doctests to all `FloatCore` methods.

### CODE REVIEW

**Comment by cuviper** [comment]
cc #32 @vks - I'd really appreciate a review on this, if only to make sure I didn't butcher your work. 😅

File: src/float.rs [hunk 1]
```diff
@@ -1493,8 +1493,8 @@ mod tests {
             assert!((FloatCore::to_radians(deg) - rad).abs() < 1e-6);
 
             let (deg, rad) = (deg as f32, rad as f32);
-            assert!((FloatCore::to_degrees(rad) - deg).abs() < 1e-6);
-            assert!((FloatCore::to_radians(deg) - rad).abs() < 1e-6);
+            assert!((FloatCore::to_degrees(rad) - deg).abs() < 1e-5);
+            assert!((FloatCore::to_radians(deg) - rad).abs() < 1e-5);
```
Comment (vks) [inline src/float.rs]: Why was this necessary?

File: src/float.rs [hunk 2]
```diff
@@ -164,14 +505,54 @@ pub trait FloatCore: Num + NumCast + Neg<Output = Self> + PartialOrd + Copy {
     }
 
     /// Returns `true` if `self` is positive, including `+0.0` and
-    /// `FloatCore::infinity()`.
+    /// `FloatCore::infinity()`, and with newer versions of Rust
+    /// even `FloatCore::nan()`.
```
Comment (vks) [inline src/float.rs]: Can we be explicit about the Rust version?

Comment (cuviper) [reply src/float.rs]: Yeah, I'll look that up.

File: src/float.rs [hunk 3]
```diff
@@ -218,11 +218,11 @@ pub trait FloatCore: Num + NumCast + Neg<Output = Self> + PartialOrd + Copy {
     #[inline]
     fn powi(mut self, mut exp: i32) -> Self {
         if exp < 0 {
-            exp = -exp;
+            exp = exp.wrapping_neg();
             self = self.recip();
         }
         // It should always be possible to convert a positive `i32` to a `usize`.
-        super::pow(self, exp.to_usize().unwrap())
+        super::pow(self, exp as u32 as usize)
```
[the remark this answers is not in the document]
Comment (cuviper) [reply src/float.rs]: To handle `i32::MIN`. See above that I switched it to `wrapping_neg`, which will leave `MIN` the same. We want that to become an unsigned positive 0x80000000, but `to_usize()` would just fail.

File: src/float.rs (hunk 1 above)
Comment (cuviper) [reply src/float.rs]: After I made it forward to std's `f32::to_degrees`, it started failing on nightly. I suspect rust-lang/rust#47919, which tried to be more accurate, actually made things *less* accurate for these inputs, but I haven't investigated yet. I will probably file an upstream bug.

**Comment by cuviper** [comment]
> What is the motivation behind adding NumCast to FloatCore? Just symmetry with Float? I think the traits are kind of orthogonal. It seems like you need NumCast for the implementation of round.

For symmetry, yes.  I thought it made sense to keep `Float` and `FloatCore` as similar as possible.  Do you see any situation where it's a *problem* to have it?

I did end up using `NumCast` for `round`, but just for a simple 0.5 -- we could avoid that with `one() / (one() + one())` if necessary.  Or we could move that implementation of `round` directly to `f32` and `f64` (`no_std` only), leaving no default.

> Should we use associated constants rather than static methods? Would that raise our minimal supported Rust version?

It would raise it a lot -- associated consts were only made stable in 1.20.

> Are the implementations of the rounding functions canonical? I don't know enough about the properties of floats to vouch for their correctness. It seems like tests for them are missing?

I implemented them independently based on their description and observed behavior.  I'm fairly confident in them, but not 100% sure.  They do have doctests where I tried to exercise interesting cases, and note this gets tested both with and without forwarding to `std`, so I know we're at least consistent for the inputs I selected.

File: src/float.rs (hunk 1 above)
Comment (cuviper) [reply src/float.rs]: https://github.com/rust-lang/rust/issues/48617

File: src/float.rs (hunk 2 above)
Comment (cuviper) [reply src/float.rs]: Changed the comment to mention that this was Rust 1.20.

File: src/float.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cuviper) [reply src/float.rs]: I compromised -- added a comment why the `as u32` is needed, but went back to using `to_usize()` so we won't be surprised here by a platform with 16-bit `usize` (although that will panic).

**Comment by cuviper** [comment]
I'm going to merge this so we can publish.  The rounding parts that I'm less confident about can always be bug-fixed later.  Thanks for reviewing!

bors r=vks
